banking-app/src/endpoints/banking.ts

```diff
@@ -0,0 +1,544 @@
+* Params:
+  * fingerprint: The certificate fingerprint of the caller.
+*/
+function formatCertFingerprint(fingerprint: string): string {
+  try{
+    fingerprint = fingerprint.toUpperCase();
+    let chars = [...fingerprint];
+    let formatted: string = fingerprint.substring(0,2);
+
+    for(let i=2;i<64;i++) {
+      if (i%2 == 0) {
+        formatted = formatted + ":" + chars[i];
+      } else {
+        formatted = formatted + chars[i];
+      }
+    }
+
+    return formatted;
+  }
+  catch (e) {
+    console.error(`Error when formatting cert fingerprint:`, e);
+    return "";
+  }
+}
+
+interface Caller {
+  id: string;
+}
+
+/**
+* Retrieve the caller id from the request. This function assumes that the endpoints 
+* are protected either using JWT or any_cert auth.
+* Returns the caller id represented as a string. 
+* Refer to https://microsoft.github.io/CCF/main/build_apps/js_app_bundle.html 
+* for the supported authentication schemes.
+* Params:
+  * request: The incoming request. 
+*/
+function getCallerId(request: ccfapp.Request<any>): string {
+  // Try to retrieve the caller cert.
+  let callerId: string = "";
+
+  const certIdentity = request.caller as unknown as ccfapp.AnyCertAuthnIdentity;
+  const callerCert = certIdentity?.cert;
+  if (callerCert) {
+    callerId = formatCertFingerprint(ccf.pemToId(callerCert));
+  } else {
+    // Try to retrieve the jwt.
+    const jwtIdentity = request.caller as unknown as ccfapp.JwtAuthnIdentity;
+    callerId = jwtIdentity?.jwt?.payload?.oid;
+  }
+  
+  if (callerId === "" || callerId === undefined) {
+    console.error("Either user cert or JWT is required")
+    throw "Either user cert or JWT is required"
+  }
+
+  console.log(`The caller id is ${callerId}`)
+  return callerId;
+}
+
+/**
+* Check if a user exist in the ledger.
+* Returns a boolean to indicate if the user exists.
+* Params:
+  * userId: The caller id from the request.
+*/
+function isValidUser(userId: string): boolean {
+    // Get a handle to the public:confidentialledger.roles.user_roles_mapping table.
+    //
+    let userRolesMapHandle = ccf.kv[userRolesMapTable];
+    if (!userRolesMapHandle) {
+        console.error(`Table: ${userRolesMapTable} does not exist`);
+        return false;
+    }
+
+    // Check if the user exist
+    //
+    const quotedUserId = `"${userId}"`;
+    return userRolesMapHandle.has(ccf.strToBuf(quotedUserId));
+}
+
+/**
+ * Check if the input is a positive integer value.
+ * Returns a boolean.
+ * Params:
+  * value: The value to be checked.
+*/
+function isPositiveInteger(value: any): boolean {
+  return Number.isInteger(value) && value > 0;
+}
+
+/**
+ * Check if the user is allowed to perform the action. 
+ * Returns a boolean to indicate if the action is allowed.
+ * Params:
+  * userId: The request caller id.
+  * action: A string that identifies the action being performed.
+*/
+function isActionAllowed(userId: string, action: string): boolean {
```

Comment:
I thought @sidmore had exposed an API that provided this functionality? It would be ideal to showcase that I think.

Reply:
I will check with him if the library is published. If it's not ready yet, I suggest that we move ahead with the current implementation and update it when the library is available.